[FX] make ASTReriter patch wrapped functions properly

mostafaelhoushi opened the pull request:
reference the same global namespace (instead of copying it) in ASTRewriter to patch wrapped functions properly

Fixes #{62071}

mostafaelhoushi committed:
make ASTReriter patch wrapped functions properly

reference the same global namespace (instead of copying it) in ASTRewriter to patch wrapped functions properly
Fixing issue: https://github.com/pytorch/pytorch/issues/62071

jamesr66a commented on torch/fx/experimental/rewriter.py:
```diff
@@ -34,7 +34,7 @@ def rewrite(self, fn: FunctionType):
         # Pull out the compiled fucntion from the newly-created Module
         code = compile(dest_ast, "", "exec")
-        globals_dict = copy.copy(fn.__globals__)
+        globals_dict = fn.__globals__
```
So instead of changing this line you want to put `fn_compiled.__globals__` down below line 42. Copying the globals here is fine and we just do it so that we can extract out the `forward`. What we want to do is to make it so that the function we ultimately return has the same globals as the input function.

mostafaelhoushi replied:
Thanks James. I have tried:
```
        globals_dict = copy.copy(fn.__globals__)
        ...
        fn_compiled.__globals__ = fn.__globals__
```

but I get a `readonly attribute error`:
```
  File "/home/melhoushi/local/miniconda3/envs/pytorch/lib/python3.9/site-packages/torch/fx/experimental/rewriter.py", line 45, in rewrite
    fn_compiled.__globals__ = fn.__globals__
AttributeError: readonly attribute
```.

Also, my original method is not working well either:
```
        globals_dict = fn.__globals__ 
        ....
```
results in the `assert len(new_keys)==1` failing for many test cases in `test_fx.py`.

mostafaelhoushi replied:
I tried to create a new class that inherits from `FunctionType` and add a `set_globals(...)` function but it seems that `FunctionType` class is set by Python to be "final" and can't be overwritten. 

Will continue to investigate other methods to solve that.

mostafaelhoushi replied:
Thanks @jamesr66a , the first method you pointed out to worked. I implemented it in this commit: e2468c7

jamesr66a reviewed (commented):
Cool! Please also add a test for this case in `test/test_fx.py`. A good source for code for the test is the reproduction code in the original issue

mostafaelhoushi committed:
Copy function with overwritten globals

Insted of passing __globals__ as reference, we copy it, and eventually copy fn_compiled with the modified globals

mostafaelhoushi committed:
test case to verify wrapped functions with ast_rewriter

mostafaelhoushi commented:
> Cool! Please also add a test for this case in `test/test_fx.py`. A good source for code for the test is the reproduction code in the original issue

I also added a test case in this commit: 7c698a0

mostafaelhoushi committed:
Remove trialing spaces to pass lint test

mostafaelhoushi committed:
Fix runtime error

When adding _copy_func, I forgot to impor functools

mostafaelhoushi committed:
remove trailing spaces to pass lint test

mostafaelhoushi committed:
remove trailing spaces to pass lint test

jamesr66a reviewed (approved):
Great! Thanks for this

cc @ansley to double check but I think you're good to land

mostafaelhoushi committed:
add other versions to verify wrapping w ast_rewriter

mostafaelhoushi commented:
I added one more commit 9817a91 to test different types of wrapping with ASTRewriter

ansley commented on test/test_fx.py:
```diff
@@ -2276,6 +2276,96 @@ def forward(self, x: torch.Tensor):
 
         traced.graph.lint()
 
+    def test_ast_rewriter_wrap(self):
```
Great tests! Thanks for being so thorough

mostafaelhoushi committed:
simplify function that copies func with globals

mostafaelhoushi commented on torch/fx/experimental/rewriter.py:
```diff
@@ -103,3 +105,20 @@ def __init__(self, orig):
         # Rewrite this single free function
         return AST_Rewriter().rewrite(cast(FunctionType, fn))
+
+def _copy_func(f, globals=None, module=None):
```
Thanks Ansley. Currently I have no plans for the helper function, but I thought making it generic maybe helpful for others in the future.

I removed the checks related to `globals` and `module` in this commit: e52bb5b. I changed the function name from `_copy_func` to `_copy_func_and_overwrite_globals` because I feel that the function is now no longer generic in copying functions. However, if you prefer changing this back to `_copy_func` I can do that.

mostafaelhoushi replied:
OK. I renamed the helper function to `change_func_globals` and reduced its scope to be within the `rewrite` function only. Hope it makes more sense now.

mostafaelhoushi committed:
rename helper function and reduce its scope
